veps/vep-4.md

```diff
@@ -0,0 +1,185 @@
+    fs.StringVar(&topoImplementation, "topo_implementation", "", "the topology implementation to use")
+    fs.StringVar(&topoGlobalServerAddress, "topo_global_server_address", "", "the address of the global topology server")
+    fs.StringVar(&topoGlobalRoot, "topo_global_root", "", "the path of the global topology data in the global topology server")
+    fs.DurationVar(&remoteOperationTimeout, "remote_operation_timeout", 30*time.Second, "time to wait for a remote operation")
+}
+
+func RemoteOperationTimeout() time.Duration {
+    return remoteOperationTimeout
+}
+```
+</details>
+<br>
+
+And then every reference to `*topo.RemoteOperationTimeout` throughout the
+codebase needs to update to `topo.RemoteOperationTimeout()`. Because these
+refactors do not affect RPC-boundaries, we can make these breaking changes at
+any point (i.e. a `vtgate` running a version of the code after the refactor and
+talk to a `vttablet` running a version of the code before the refactor without
+issue). In addition, the various `./go/cmd/<component>/main.go` files would
+need to audit whether they depend on these flags, and be sure to add a call to
+`topo.AddFlags(flag.CommandLine)`, which would add the newly-scoped
+`topo`-related flags back into the global flagset, but now it's explicit rather
+than implicit.
+
+I'll also point out that `./go/vt/servenv` uses this pattern specifically for
+the `-socket_file` flag, but none of the other flags in that package.
+
+### Tradeoffs
+
+* ➕ Pro: This refactoring form allows us to work through each packages' flags incrementally.
+  We can refactor a single package, and audit any binaries that may rely on it as a single, contained PR; it's not an all-at-once/all-or-nothing refactor.
+* ➕ Pro: Making all flags package-private gives the package that "owns" the flag complete control
+  over how that flag is accessed (e.g. could guard a flag behind a mutex if we wanted to).
+  * ➕ Pro: Having complete control over the access gives us the ability to swap out access "implementations"
+  transparently to callers, paving the way for us to drop in libraries like [viper] in cases where it's appropriate, but without making that a hard requirement.
+  * ➖ Con: There are currently certain places where we mutate another package's flag value (mostly in tests, but we would need to do an audit of this). "Package-private variable + getter function" won't fit this model, so we would need to refactor that code, or provide setter functions in some cases.
+* ➖ Con: We need to specify the default values for each flag twice. Once, where the
+  variable is declared, to cover the case where the binary never made a call to
+  `AddFlags()` and we should implicitly rely on the default value. Then, a
+  duplication of the default value must also appear within `AddFlags()` to pass
+  to the relevant `flag.<Type>Var(...)` call.
+
+### Subpackages
+
+(NB: I'm not convinced this is a good idea at this point, but I wanted to put
+it out there for criticism).
```

Comment:
Not sure how much complexity this adds, but this seems like a very solid goal to me, given that in vitessio/vitess#7471 we're looking at having a server and a client binary.